pkg/util/context.go

```diff
@@ -0,0 +1,22 @@
+package util
+
+import "context"
+
+// ContextForChannel derives a child context from a parent channel.
+//
+// The derived context's Done channel is closed when the returned cancel function
+// is called or when the parent channel is closed, whichever happens first.
+//
+// Note the caller must *always* call the CancelFunc, otherwise resources may be leaked.
+func ContextForChannel(parentCh <-chan struct{}) (context.Context, context.CancelFunc) {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	go func() {
+		select {
+		case <-parentCh:
+			cancel()
+		case <-ctx.Done():
+		}
+	}()
```

Comment:
My concern is here will start a goroutine for each cluster, and the goroutine is keeping detects the two channels(different with `blocked`), not sure if there is a performance risk if we have a bunch of clusters.

I'll investigate it a little bit.

Reply:
Sorry, `select` will `bloecked` until any of the cases is ready to operate. So, no performance issue here.

Reply:
Maybe we can use `context` instead of `channel` for both `NewMultiClusterInformerManager` and `NewSingleClusterInformerManager`.  In that case ,we can save the goroutine. 

We can do it after this PR.

Reply:
Good idea.